crates/ra_syntax/src/algo.rs

```diff
@@ -123,6 +124,37 @@ pub fn replace_children(
     with_children(parent, new_children)
 }
 
+/// Replaces descendants in the node, according to the mapping.
+///
+/// This is a type-unsafe low-level editing API, if you need to use it, prefer
+/// to create a type-safe abstraction on top of it instead.
+pub fn replace_descendants(
+    parent: &SyntaxNode,
+    map: &FxHashMap<SyntaxElement, SyntaxElement>,
```

Comment:
Why not take a closure mapping from `SyntaxElement`  to `Option<SyntaxElement>` instead?

Reply:
Ideally, we would “zip” all the keys of the map with the original subtree, in order to avoid descending into not-modified branches. For this, you need to know the whole set of keys beforehand